### PR TITLE
Create specifications from `@NonNull` and `@Pure` annotations

### PR DESCRIPTION
**Pull request by RockingMat**
Included nonull annotations in building specifications and pure annotated methods were included as side-effect-free

### CODE REVIEW

File: src/main/java/randoop/operation/TypedOperation.java [hunk 1]
```diff
@@ -69,6 +84,111 @@ public abstract class TypedOperation implements Operation, Comparable<TypedOpera
...
+    List<Precondition> preconditions = new ArrayList<>();
+    List<Postcondition> postconditions = new ArrayList<>();
+    List<ThrowsCondition> throwsConditions = new ArrayList<>();
+
+    // Read parameter annotations
+    Parameter[] parameters = method.getParameters();
+    for (Parameter parameter : parameters) {
+      AnnotatedType annotatedType = parameter.getAnnotatedType();
+      for (Annotation annotation : annotatedType.getAnnotations()) {
+        if (annotation.annotationType().getSimpleName().equals("NonNull")) {
```
Comment (mernst) [inline src/main/java/randoop/operation/TypedOperation.java:135]: Please also check for "NotNull".

**Review by mernst — CHANGES_REQUESTED** [review]

File: src/main/java/randoop/operation/TypedOperation.java (hunk 1 above)
Comment (RockingMat) [reply src/main/java/randoop/operation/TypedOperation.java]: @mernst I changed the HashSets to LinkedHashSets. Let me know if there are any other changes that would be useful.

File: src/main/java/randoop/operation/TypedOperation.java [hunk 2]
```diff
@@ -70,6 +84,83 @@ public abstract class TypedOperation implements Operation, Comparable<TypedOpera
...
+      Method method, @ClassGetName String className) {
+    String methodName = method.getName();
+    List<String> parameterNames =
+        CollectionsPlume.mapList(Parameter::getName, method.getParameters());
+    List<@ClassGetName String> parameterTypes =
+        CollectionsPlume.mapList(Class::getName, method.getParameterTypes());
+
+    OperationSignature operation =
+        OperationSignature.forConstructorName(className, methodName, parameterTypes);
+    Identifiers identifiers = new Identifiers("receiver", parameterNames, "result");
```
Comment (mernst) [inline src/main/java/randoop/operation/TypedOperation.java:126]: Is a `result` identifier always needed?  What if the method has `void` return type?

**Review by mernst — CHANGES_REQUESTED** [review]
I have one concern; maybe adding a test regarding it would  be a good idea.

File: src/main/java/randoop/operation/TypedOperation.java (hunk 2 above)
Comment (RockingMat) [reply src/main/java/randoop/operation/TypedOperation.java:126]: The Identifiers class should assign a "result" identifier in it's constructor regardless of the return type. If the return type is void I do not think the method would have a pure annotation.

**Pull request merged**